### PR TITLE
feat(ext/node): buffer.transcode()

### PR DESCRIPTION
**Pull request by satyarohith**
Closes https://github.com/denoland/deno/issues/25911

### CODE REVIEW

File: ext/node/polyfills/internal/buffer.mjs [hunk 1]
```diff
@@ -2598,6 +2598,31 @@ export function isAscii(input) {
...
+    fromEnco === "ascii" && toEnco === "latin1" ||
+    fromEnco === "ucs2" && toEnco === "utf16le" ||
+    fromEnco === "utf16le" && toEnco === "ucs2";
+  if (returnSource) {
+    return Buffer.from(source);
+  }
+
+  // TODO(@satyarohith): return appropriate error codes when transcode fails.
+  // The fix should enable the commented tests in test-icu-transcode.js
+  const result = op_transcode(new Uint8Array(source), fromEnco, toEnco);
```
Comment (bartlomieju) [inline ext/node/polyfills/internal/buffer.mjs:2622]: Can this be addressed before we land?

Comment (satyarohith) [reply ext/node/polyfills/internal/buffer.mjs]: Yes, addressed.

File: tests/node_compat/test/parallel/test-icu-transcode.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+// assert.throws(
+//   () => buffer.transcode(Buffer.from('a'), 'b', 'utf8'),
+//   /^Error: Unable to transcode Buffer \[U_ILLEGAL_ARGUMENT_ERROR\]/
+// );
+
+// assert.throws(
+//   () => buffer.transcode(Buffer.from('a'), 'uf8', 'b'),
+//   /^Error: Unable to transcode Buffer \[U_ILLEGAL_ARGUMENT_ERROR\]$/
+// );
```
Comment (satyarohith) [inline tests/node_compat/test/parallel/test-icu-transcode.js:69]:
```suggestion
assert.throws(
  () => buffer.transcode(Buffer.from('a'), 'b', 'utf8'),
  /^Error: Unable to transcode Buffer \[U_ILLEGAL_ARGUMENT_ERROR\]/
);

assert.throws(
  () => buffer.transcode(Buffer.from('a'), 'uf8', 'b'),
  /^Error: Unable to transcode Buffer \[U_ILLEGAL_ARGUMENT_ERROR\]$/
);
```

File: ext/node/ops/buffer.rs [hunk 3]
```diff
@@ -11,3 +13,108 @@ pub fn op_is_ascii(#[buffer] buf: &[u8]) -> bool {
...
+
+fn utf8_to_ucs2(utf8_bytes: &[u8]) -> Result<Vec<u8>> {
+  let utf8_string = std::str::from_utf8(utf8_bytes)?;
+  let ucs2_vec: Vec<u16> = utf8_string.encode_utf16().collect();
+  let bytes: Vec<u8> = ucs2_vec.iter().flat_map(|&x| x.to_le_bytes()).collect();
+  Ok(bytes)
+}
+
+fn utf8_to_latin1(utf8_bytes: &[u8]) -> Vec<u8> {
+  let mut latin1_bytes = Vec::new();
```
Comment (littledivy) [inline ext/node/ops/buffer.rs:64]:
```suggestion
  let mut latin1_bytes = Vec::with_capacity(utf8_bytes.len());
```

File: ext/node/ops/buffer.rs [hunk 4]
```diff
@@ -11,3 +13,108 @@ pub fn op_is_ascii(#[buffer] buf: &[u8]) -> bool {
...
+          i += 1;
+        }
+      }
+    }
+  }
+  latin1_bytes
+}
+
+fn utf8_to_ascii(source: &[u8]) -> Vec<u8> {
+  let mut ascii_bytes = Vec::new();
```
Comment (littledivy) [inline ext/node/ops/buffer.rs:99]:
```suggestion
  let mut ascii_bytes = Vec::with_capacity(source.len());
```

File: ext/node/polyfills/internal/buffer.mjs [hunk 5]
```diff
@@ -2598,6 +2602,40 @@ export function isAscii(input) {
...
+  }
+  fromEnco = fromEnco === "binary" ? "latin1" : fromEnco;
+  toEnco = toEnco === "binary" ? "latin1" : toEnco;
+  const returnSource = fromEnco === "ascii" && toEnco === "utf8" ||
+    fromEnco === "ascii" && toEnco === "latin1" ||
+    fromEnco === "ucs2" && toEnco === "utf16le" ||
+    fromEnco === "utf16le" && toEnco === "ucs2";
+  if (returnSource) {
+    return Buffer.from(source);
+  }
```
Comment (littledivy) [inline ext/node/polyfills/internal/buffer.mjs:2624]: Please add a comment about what's going on here

File: ext/node/polyfills/internal/buffer.mjs [hunk 6]
```diff
@@ -2598,6 +2602,40 @@ export function isAscii(input) {
...
+      "source",
+      ["Buffer", "Uint8Array"],
+      source,
+    );
+  }
+  if (source.length === 0) {
+    return Buffer.alloc(0);
+  }
+  fromEnco = fromEnco === "binary" ? "latin1" : fromEnco;
+  toEnco = toEnco === "binary" ? "latin1" : toEnco;
```
Comment (littledivy) [inline ext/node/polyfills/internal/buffer.mjs:2617]: We may want to call [`normalizeEncoding`](https://github.com/denoland/deno/blob/3881b7173445ab7f68ec94d5dedbb1cb1c1978ec/ext/node/polyfills/_utils.ts#L101-L106) here. It takes care of casing.

**Review by littledivy — APPROVED** [review]
LGTM

**Pull request merged**